src/main/java/bio/terra/stairway/FlightDao.java

```diff
@@ -1,301 +1,371 @@
+                if (flightStateList.size() > 1) {
+                    throw new DatabaseOperationException("Multiple flights with the same id?!");
+                }
+                return flightStateList.get(0);
+            }
 
-        try {
-            return jdbcTemplate.queryForObject(sqlOneFlight, params, new FlightStateMapper());
-        } catch (EmptyResultDataAccessException emptyEx) {
-            throw new FlightNotFoundException("Flight not found: " + flightId, emptyEx);
-        } catch (IncorrectResultSizeDataAccessException sizeEx) {
-            throw new DatabaseOperationException("Multiple flights with the same id?! " + flightId, sizeEx);
+        } catch (SQLException ex) {
+            throw new DatabaseOperationException("Failed to get flight", ex);
         }
     }
 
-    public List<FlightState> getFlights(int offset, int limit) {
+    public List<FlightState> getFlights(int offset, int limit) throws DatabaseOperationException {
         return getFlights(offset, limit, "", Collections.EMPTY_MAP);
     }
 
-    public List<FlightState> getFlightsForUser(int offset, int limit, String subject) {
+    List<FlightState> getFlightsForUser(int offset, int limit, String subject) throws DatabaseOperationException {
         return getFlights(offset, limit, " WHERE owner_id = :subject ", Collections.singletonMap("subject", subject));
     }
 
-    private List<FlightState> getFlights(int offset, int limit, String whereClause, Map<String, ?> whereParams) {
+    private List<FlightState> getFlights(int offset, int limit, String whereClause, Map<String, String> whereParams)
+        throws DatabaseOperationException {
+
         final String sqlFlightRange = "SELECT flightid, submit_time, input_parameters," +
-            " completed_time, output_parameters, status, exception, owner_id, owner_email" +
+            " completed_time, output_parameters, status, serialized_exception, owner_id, owner_email" +
             " FROM " + FLIGHT_TABLE +
             whereClause +
             " ORDER BY submit_time" +  // should this be descending?
             " LIMIT :limit OFFSET :offset";
 
-        MapSqlParameterSource params = new MapSqlParameterSource()
-            .addValue("limit", limit)
-            .addValue("offset", offset)
-            .addValues(whereParams);
+        try (Connection connection = dataSource.getConnection();
+             NamedParameterPreparedStatement flightRangeStatement =
+                 new NamedParameterPreparedStatement(connection, sqlFlightRange)) {
 
-        List<FlightState> flightStateList = jdbcTemplate.query(sqlFlightRange, params, new FlightStateMapper());
-        return flightStateList;
+            flightRangeStatement.setInt("limit", limit);
+            flightRangeStatement.setInt("offset", offset);
+            for (Map.Entry<String, String> entry : whereParams.entrySet()) {
+                flightRangeStatement.setString(entry.getKey(), entry.getValue());
+            }
```

Comment:
This for loop seems like a good candidate to add to the NamedParameterPreparedStatement object -- accept a Map instead. Maybe not as part of this PR, but perhaps a TODO for the future?

Reply:
I added a TODO. It would be nicer to do something more generic than Map<String, String>, for a general feature.